sway-ir/src/verify.rs

```diff
@@ -548,11 +548,14 @@ impl InstructionVerifier<'_, '_> {
                     return Err(IrError::VerifyBinaryOpIncorrectArgType);
                 }
             }
-            BinaryOpKind::Add
-            | BinaryOpKind::Sub
-            | BinaryOpKind::Mul
-            | BinaryOpKind::Div
-            | BinaryOpKind::Mod => {
+            BinaryOpKind::Add => {
+                if !(arg1_ty.eq(self.context, &arg2_ty) && arg1_ty.is_uint(self.context)
+                    || arg1_ty.is_ptr(self.context) && arg2_ty.is_uint64(self.context))
+                {
+                    return Err(IrError::VerifyBinaryOpIncorrectArgType);
+                }
+            }
+            BinaryOpKind::Sub | BinaryOpKind::Mul | BinaryOpKind::Div | BinaryOpKind::Mod => {
```

Comment:
Shouldn't the same logic as above be valid for `Sub`? We can also subtract from pointers. In fact, we have the `__ptr_add` and `__ptr_sub` intrinsics that use the `Add` and `Sub` respectively. There is only one test for `__ptr_sub` and I assume that one is passing because we still treat `raw_ptr` as `u64` so the `Sub` ends up having `u64` and `u64` as arguments. But once we represent `raw_ptr` with `IrType::Ptr` as we do references in this PR, I assume the `__ptr_sub` will fail because the `Sub` as is will not accept `Ptr` as the first parameter.

Reply:
You are correct.

But should I make this change now is the question. Perhaps it's better to have a case possible (as you describe the `raw_ptr` case) where it'll fail, and then we make the change? We'll have a unit test too then. What do you think? I don't have a strong opinion though, so can make the change now.

Reply:
Representing `raw_ptr` as IR pointer will anyhow be one of the next steps, so I agree, the change can wait until then.